docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+  key range.  However, such key range settings are currently ignored in FIFO
+  Compaction as RocksDB assumes SST files under FIFO Compaction never have
+  disjoint key ranges.  A change which adds additional option to CompactionOptions
+  to force applying key range to CompactRange to FIFO Compaction.
+
+- CompactFile, on the other hand, directly runs compaction on the specified
+  SST file.  If we want to specifically compact one SST file, a PR is needed
+  to make this API available in RocksDB's Rust binding.
+
+#### Avoid Using the Optional FIFO Background Compaction
+One thing to mention is that FIFO Compaction does offer a way to enable
+automatic background compaction via an option called
+[CompactionOptionsFIFO.allow_compaction](
+https://github.com/facebook/rocksdb/wiki/FIFO-compaction-style#compaction),
+which allows FIFO Compaction have trigger background compaction when the number
+of files exceeds the configured threshold.  However, it is not suitable for
+Solana's use case for the following two main reasons:
+
+ - The background compactions triggered by this setting ignores the fact that
+   writes to Solana's BlockStore are mostly sorted.  As a result, the background
+   compactions tend to merge as many files as possible and produce one big
+   file, and the files chosen by such compaction might not necessary pick
+   the set of files we actually want to compact.
+ 
+ - The resulting output file will have newer time stamp while it actually
+   contains older files, which over time makes FIFO Compaction throw away
+   files with random time order.
+
+#### Compaction Filters are Skipped By Default
+Because FIFO Compaction does not support background compaction by default,
+all compaction filters are ignored unless manual compactions are issued.
+
+Enabling manual compactions that suitable for Solana's BlockStore is
+mentioned in the earlier section.
+
+#### Deletions via Delete() API Must be Done Carefully
+Deletions via RocksDB's Delete API is discouraged in FIFO Compaction for
+two reasons.
+
+First, FIFO Compaction does not run background compaction. As a result,
+deleted entries will only become invisible to the read side, and the
+deleted entries will continue occupying disk storage until its belonging
+SST file is old enough to pass the FIFO Compaction TTL setting.
+
+Second, deletions via Delete() API are likely to destroy the monotonic
+increasing key order property.  Specifically, we should avoid deleting
+older keys via the Delete() API.  If necessary, massive deletions must
+be followed by a flush in order to have a SST file which key range does
+not overlap too much with the other SST files.
+
+#### Test Coverage
+While FIFO Compaction is one of the three compaction styles offered by
+RocksDB, it is not the primary compaction styles and thus have relative less
+test coverage.  Additional test cases or Solana specific test cases might
+be needed to have better coverage.
```

Comment:
good point. that's my 2nd concern to switching to non-default fifo, combined with my general lack of familarity with rocksdb (esp, fifo). i think we can now send patches to upstream when using fifo? :)

Reply:
Yep, if the benchmark result for FIFO on Solana workload looks promising, I can continue working on all the necessary patches but looks like the RPC side might have higher priority than this.